### PR TITLE
add discovery hook to speed up node finding for server

### PR DESCRIPTION
**Pull request by lucusfly**
This change is [Reviewable](https://reviewable.io/reviews/seeleteam/go-seele/89)

### CODE REVIEW

**Comment by boqiu** [comment]
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[p2p/message.go, line 32 at r1](https://reviewable.io:443/reviews/seeleteam/go-seele/89#-L8AkhK2Ykv2H6o4qj0g:-L8AkhK2Ykv2H6o4qj0h:b3s80lj) ([raw file](https://github.com/seeleteam/go-seele/blob/760a4505a72e0cc6f91a5ca56d77149e25a5ddbb/p2p/message.go#L32)):*
> ```Go
> type ProtoHandShake struct {
> 	Caps   []Cap
> 	NodeID common.Address
> ```

@qiaozhigang please double confirm. Once we change the Address (e.g. ETH), any impact?

---

*[p2p/discovery/database.go, line 21 at r1](https://reviewable.io:443/reviews/seeleteam/go-seele/89#-L8AkFn5xil9GrSzPFN-:-L8AkFn5xil9GrSzPFN0:b-r91sf3) ([raw file](https://github.com/seeleteam/go-seele/blob/760a4505a72e0cc6f91a5ca56d77149e25a5ddbb/p2p/discovery/database.go#L21)):*
> ```Go
> 	mutex          sync.Mutex
> 	addNodeHook    NodeHook
> 	deleteNodeHook NodeHook
> ```

Basically, it's OK in this PR. But suggest to use existing event framework to replace the hook way.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/seeleteam/go-seele/89)*

**Comment by WTer** [comment]
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[p2p/discovery/database.go, line 90 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/89#-L8BKMsS8BJD3KwuHn8f:-L8BKMsS8BJD3KwuHn8g:b-fyhjvq) ([raw file](https://github.com/seeleteam/go-seele/blob/760a4505a72e0cc6f91a5ca56d77149e25a5ddbb/p2p/discovery/database.go#L90)):*
> SetHookForDeleteNode

no any reference to this function?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/seeleteam/go-seele/89)*

**Comment by lucusfly** [comment]
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[p2p/discovery/database.go, line 21 at r1](https://reviewable.io:443/reviews/seeleteam/go-seele/89#-L8AkFn5xil9GrSzPFN-:-L8BQcq16YH2MTuwO2QB:b238mlm) ([raw file](https://github.com/seeleteam/go-seele/blob/760a4505a72e0cc6f91a5ca56d77149e25a5ddbb/p2p/discovery/database.go#L21)):*
<details><summary><i>Previously, boqiu wrote…</i></summary><blockquote>

Basically, it's OK in this PR. But suggest to use existing event framework to replace the hook way.
</blockquote></details>

For some concerns, I don't want too many  packages use and depend on this hook. so just keep it.

---

*[p2p/discovery/database.go, line 90 at r1](https://reviewable.io:443/reviews/seeleteam/go-seele/89#-L8BKMsS8BJD3KwuHn8f:-L8BQ9cDNQlolFpJkw3u:bxqiw1) ([raw file](https://github.com/seeleteam/go-seele/blob/760a4505a72e0cc6f91a5ca56d77149e25a5ddbb/p2p/discovery/database.go#L90)):*
<details><summary><i>Previously, WTer wrote…</i></summary><blockquote>

> SetHookForDeleteNode

no any reference to this function?
</blockquote></details>

actually, server use ping pong to detect inactive node. so it may not be necessary to use this hook. But we need to think about it later.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/seeleteam/go-seele/89)*

**Comment by lucusfly** [comment]
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[p2p/message.go, line 32 at r1](https://reviewable.io:443/reviews/seeleteam/go-seele/89#-L8AkhK2Ykv2H6o4qj0g:-L8BQxngR1sqAsFgAOwY:b-t3tz34) ([raw file](https://github.com/seeleteam/go-seele/blob/760a4505a72e0cc6f91a5ca56d77149e25a5ddbb/p2p/message.go#L32)):*
<details><summary><i>Previously, boqiu wrote…</i></summary><blockquote>

@qiaozhigang please double confirm. Once we change the Address (e.g. ETH), any impact?
</blockquote></details>

NodeID and Address are actually same type.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/seeleteam/go-seele/89)*

**Comment by WTer** [comment]
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/seeleteam/go-seele/89)*

**Comment by qiaozhigang** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[p2p/message.go, line 32 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/89#-L8AkhK2Ykv2H6o4qj0g:-L8BbNW37vNFGmvJtMvX:b-pbw7ax) ([raw file](https://github.com/seeleteam/go-seele/blob/760a4505a72e0cc6f91a5ca56d77149e25a5ddbb/p2p/message.go#L32)):*
<details><summary><i>Previously, lucusfly (lucus) wrote…</i></summary><blockquote>

NodeID and Address are actually same type.
</blockquote></details>

Should be type nodeid, but now it is same. 
It must be refactored when we redifine common.address in future.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/seeleteam/go-seele/89)*

**Comment by qiaozhigang** [comment]
Reviewed 1 of 5 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/seeleteam/go-seele/89)*

**Comment by boqiu** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[p2p/message.go, line 32 at r1](https://beta.reviewable.io/reviews/seeleteam/go-seele/89#-L8AkhK2Ykv2H6o4qj0g:-L8BcZ8T7IzZcWx0Mk2t:b-tdvmjc) ([raw file](https://github.com/seeleteam/go-seele/blob/760a4505a72e0cc6f91a5ca56d77149e25a5ddbb/p2p/message.go#L32)):*
<details><summary><i>Previously, qiaozhigang wrote…</i></summary><blockquote>

Should be type nodeid, but now it is same. 
It must be refactored when we redifine common.address in future.
</blockquote></details>

Yes, that's what I concerned. If we change the address to ETH alike 20 bytes, then NodeID and address type are different.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/seeleteam/go-seele/89)*
